Feature/users registrer

jeantech-zz opened the pull request:
Created register user

jeantech-zz committed:
registre-user

jeantech-zz committed:
se realiza refactor de controlador para usar actiones

jeantech-zz committed:
Request validando el formulario para crear usuario

jeantech-zz committed:
refactor de test register

jeantech-zz committed:
view and fix test register

jeantech-zz committed:
refactor created user

jeantech-zz committed:
create test register

jortiz3109 commented:
OJO con las convenciones: https://www.conventionalcommits.org/en/v1.0.0/

jortiz3109 commented on app/Actions/Register/CreateUser.php:
```diff
@@ -0,0 +1,26 @@
+
+class CreateUser
+{
+    public static function execute(array $userRequest): User
```
¿Realmente es un "request" lo que estamos recibiendo como parámetro? ¿Porque es un request?

jortiz3109 commented on app/Actions/Register/CreateUser.php:
```diff
@@ -0,0 +1,26 @@
+            'updated_by' =>   auth()->id(),
+        ]);
+
+        event(new Registered($user));
```
¿Un evento si debería ser despachado desde una acción?
Es decir: ¿Realmente es reponsabilidad de la acción de crear; despachar el evento en cuestión?

jortiz3109 commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,43 @@
+
+    public function create()
+    {
+        $userId=auth()->id();
```
Hace falta el tipo de dato de retorno.
OJO con la indentación y los espacios según PSR...

jortiz3109 commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,43 @@
+    public function create()
+    {
+        $userId=auth()->id();
+        return view('register.create', [
```
¿Porque no estamos usando un ViewModel?

jortiz3109 commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,43 @@
+        ]);
+    }
+
+    public function store(UserCreateRequest $request)
```
¿Donde está el tipo de dato de retorno?
¿Porque no estamos usando un viewmodel?

jortiz3109 commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,43 @@
+    {
+        CreateUserAction::execute($request->validated());
+
+        return view('dashboard', [
```
Yo no debería retornar una vista en este método del controlador, en su lugar debería redireccionar a una ruta existente con un mensaje.

jortiz3109 commented on app/Http/Requests/Register/UserCreateRequest.php:
```diff
@@ -0,0 +1,53 @@
+
+class UserCreateRequest extends FormRequest
+{
+    /**
```
¿Estos comentarios si aportan valor?

jortiz3109 commented on app/Http/Requests/Register/UserCreateRequest.php:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @return array
+     */
+    public function rules()
```
¿Donde está el tipo de dato de retorno?

jortiz3109 commented on app/Http/Requests/Register/UserCreateRequest.php:
```diff
@@ -0,0 +1,53 @@
+        ];
+    }
+
+    public function messages()
```
¿Donde está el tipo de dato de retorno?
¿Que pasó con las traducciones?
¿Si requiero traducir los mensajes de error este si será el mejor enfoque?

jortiz3109 commented on app/Http/Requests/Register/UserCreateRequest.php:
```diff
@@ -0,0 +1,53 @@
+        ];
+    }
+
+    public function attributes()
```
¿Donde está el tipo de dato de retorno?
OJO que este no es el mejor mecanismo para traducir los nombres de los campos

jortiz3109 commented on database/migrations/2021_10_24_224610_add_column_user.php:
```diff
@@ -0,0 +1,34 @@
+     *
+     * @return void
+     */
+    public function up()
```
¿Tipo de dato de retorno?

jortiz3109 commented on database/migrations/2021_10_24_224610_add_column_user.php:
```diff
@@ -0,0 +1,34 @@
+<?php
```
¿Este nombre de la migración si representa los cambios que se están haciendo en la tabla de la DB?

jortiz3109 commented on database/migrations/2021_10_24_224610_add_column_user.php:
```diff
@@ -0,0 +1,34 @@
+     *
+     * @return void
+     */
+    public function down()
```
¿donde está el tipo de dato de retorno?
¿Que pasa con los comentarios que no aportan valor?

jortiz3109 commented on database/migrations/2021_10_24_224610_add_column_user.php:
```diff
@@ -0,0 +1,34 @@
+    public function down()
+    {
+        Schema::table('users', function (Blueprint $table) {
+            $table->dropForeign(['created_by']);
```
Estoy eliminando el constraint pero no eliminando el campo

jortiz3109 commented on database/seeders/DatabaseSeeder.php:
```diff
@@ -14,5 +14,6 @@ class DatabaseSeeder extends Seeder
     public function run()
     {
         // \App\Models\User::factory(10)->create();
+        $this->call(UserSeeder::class);
```
¿Porque es necesario este seeder?

jeantech-zz replied:
Si porque en el sistema debe existir un usuario para poder  registrar usuarios

jortiz3109 commented on database/seeders/UserSeeder.php:
```diff
@@ -0,0 +1,33 @@
+     *
+     * @return void
+     */
+    public function run()
```
Tipo de dato de retorno?
Comentarios que no aportan valor

jortiz3109 replied:
Valor de retorno?

jortiz3109 commented on database/seeders/UserSeeder.php:
```diff
@@ -0,0 +1,33 @@
+            "name"=>"Admin",
+            "email"=>"jeante05@gmail.com",
+            'email_verified_at' => now(),
+            "password"=>Hash::make('jeante05'),
```
IMPORTANTE!!! por seguridad ningún código fuente puede tener, usuarios, contraseñas u otros tipos de mecanismos de autentiación en blando.

¿Que ocurre si quiero crear un administrador con otro usuario/contraseña?

jortiz3109 commented on resources/views/layouts/admin.blade.php:
```diff
@@ -26,7 +26,7 @@
                         <ul class="menu-list">
                             <li><a href="{{ route('permissions.index') }}"><em class="pr-2 mdi mdi-shield-lock"></em>@lang('permissions.navbar.title')</a></li>
-                            <li><a href="#"><em class="pr-2 mdi mdi-account-multiple"></em>@lang('Users')</a></li>
+                            <li><a href="{{ route('register.create') }}"><em class="pr-2 mdi mdi-account-multiple"></em>@lang('Users')</a></li>
```
Creo que este enlace no va en esta vista

jeantech-zz replied:
Coloque en enlace para poder que desde el menu se pueda ingresar a registrar a usuarios desde el dashboard, como vi que esta el de permisos.

En donde debería de ir?

jortiz3109 commented on routes/web.php:
```diff
@@ -17,6 +18,9 @@
 
     Route::name('logins.index')->get('/logins', LoginLogController::class);
+
+    Route::name('register.create')->get('/register', [UserController::class, 'create']);
```
¿Porque una ruta register usa un método del controlador UserController?
Creo que el nombre no refleja la función del controlador

jortiz3109 replied:
¿Que ocurre si un usuario con la sesión iniciada accede a esta ruta?
Esta ruta creo que debería ser únicamente para usuarios invitados (guests)

jeantech-zz replied:
Segun el requerimiento para crear usuario debe hacerlo un usuario admin ya que 
"El sistema debe guardar el id del administrador que creó el usuario"

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+{
+    use RefreshDatabase;
+
+    public function test_registration_screen_can_be_rendered()
```
¿Tipo de dato de retorno?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+        $response->assertStatus(200);
+    }
+
+    public function test_new_users_can_register()
```
¿Tipo de dato de retorno?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+
+    }
+
+    public function test_required_name_user_register()
```
¿Tipo de dato de retorno?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+
+    }
+
+    public function test_required_email_user_register()
```
¿Tipo de dato de retorno?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+
+    }
+
+    public function test_uniqued_email_user_register()
```
¿Tipo de dato de retorno?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+
+    }
+
+    public function test_valid_name_user_register()
```
¿Tipo de dato de retorno?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+
+    }
+
+    public function test_disabled_user_register()
```
¿Tipo de dato de retorno?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+        
+    }
+
+    public function test_create_date_user_register()
```
¿Tipo de dato de retorno?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+        
+    }
+
+    public function test_create_id_user_register()
```
¿Tipo de dato de retorno?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,153 @@
+<?php
```
¿Que pasó con los data-providers?
Hay muchos datos repetidos que pueden unificarse

jortiz3109 reviewed (changes requested):
Hay que tener mucho cuidado con los tipos de dato de retorno.
Hay que mirar con detenimiento que comentarios dejamos en el código

jeantech-zz committed:
fix user action controller

jeantech-zz committed:
fix create register user

jeantech-zz committed:
fix: type return

jeantech-zz committed:
fix: register user

jeantech-zz committed:
fix: add create viewModel

jeantech-zz committed:
fix: comment request user

jeantech-zz committed:
fix: comment request user

jeantech-zz committed:
fix: add lang create user

jeantech-zz committed:
fix: migration add created and update user

jeantech-zz committed:
fix: migration delete constrain

jeantech-zz committed:
fix: seed type return

jeantech-zz committed:
fix: route user

jeantech-zz committed:
fix: migrate add create update user

jeantech-zz committed:
fix:test_user_register

jeantech-zz committed:
fix: delete link  create user layout

jeantech-zz committed:
fix test and controller user

jortiz3109 commented on app/Actions/Register/CreateUserAction.php:
```diff
@@ -0,0 +1,21 @@
+
+class CreateUserAction
+{
+    public static function execute(array $user): User
```
¿Este array si debería llamarse user?

jortiz3109 commented on app/Actions/Register/CreateUserAction.php:
```diff
@@ -0,0 +1,21 @@
+{
+    public static function execute(array $user): User
+    {
+        return auth()->user()->create([
```
Yo la crearía así:
```php
return User::create([
            'name' =>  $user['name'],
            'email' => $user['email'] ,
            'password' => Hash::make($user['password']),
            'created_by' =>  auth()->id(),
            'updated_by' =>   auth()->id(),
        ]);
```

jortiz3109 commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,32 @@
+    {
+        $viewModel->collection($users);
+
+        return view('register.create', $viewModel->toArray());
```
Esta vista debería llamarse users.create

jortiz3109 commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,32 @@
+
+    public function create(CreateViewModel $viewModel): View
+    {
+        $viewModel->collection($users);
```
No entiendo muy bien como funciona este viewModel

jortiz3109 commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,32 @@
+
+    }
+
+    public function store(UserCreateRequest $request)
```
¿Cual es el tipo de dato que retorna esta función?

jortiz3109 commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,32 @@
+
+        event(new Registered($user));
+
+        return redirect()->route('dashboard')->with('success', 'Register created successfully.');
```
¿Y las traducciones?

jortiz3109 commented on app/Http/Requests/Register/UserCreateRequest.php:
```diff
@@ -0,0 +1,23 @@
+    {
+        return [
+            'name' => 'required|string|max:255|regex:/^[a-zA-Z\s]+$/u',
+            'email' => 'required|string|email|max:255|unique:users',
```
La regla unique te funciona, pero creo que es mas idóneo usarlo así:

```php
Rule::unique('users')->ignore($user->id),
```

jortiz3109 commented on app/ViewModels/Users/CreateViewModel.php:
```diff
@@ -0,0 +1,33 @@
+{
+    use HasCollection;
+
+    protected function buttons(): array
```
¿No hay botones para la vista de creación?

jortiz3109 commented on app/ViewModels/Users/CreateViewModel.php:
```diff
@@ -0,0 +1,33 @@
+        return trans('users.titles.create');
+    }
+
+    public function filters(): array
```
Hay que revisar la pertinencia de usar los métodos filters() y data()

jortiz3109 commented on database/migrations/2021_10_24_224610_add_create_update_user.php:
```diff
@@ -0,0 +1,24 @@
+    public function up(): void
+    {
+        Schema::table('users', function (Blueprint $table) {
+            $table->foreignId('created_by')->nullable()->constrained('users')->onDelete('cascade');
```
Ojo con el onDelete de estos constraints, porque pueden derivar en eliminaciones no intencionales de datos.

jortiz3109 commented on database/seeders/UserSeeder.php:
```diff
@@ -0,0 +1,33 @@
+    public function run()
+    {
+       
+        $user =User::create([
```
Este código no cumple con PSR

jortiz3109 commented on database/seeders/UserSeeder.php:
```diff
@@ -0,0 +1,33 @@
+            'enabled_at' => now(),
+        ]);
+
+        event(new Registered($user));
```
¿Este evento si es necesario?

juancolo replied:
corregido

jortiz3109 commented on resources/views/register/create.blade.php:
```diff
@@ -0,0 +1,28 @@
+@extends('layouts.admin')
```
Hay que renombrar esta vista

jortiz3109 commented on resources/views/register/create.blade.php:
```diff
@@ -0,0 +1,28 @@
+                </b-field>
+
+                <button type="submit" class="button is-primary is-fullwidth"> <b-icon pack="fas" icon="save"></b-icon>&nbsp;
+                    @lang('users.buttons.register')
```
La traducción no debería ser users.buttons.register, sino que podría ser algo como: common.buttons.save

jortiz3109 commented on resources/views/register/create.blade.php:
```diff
@@ -0,0 +1,28 @@
+    <form action="{{ route('users.store') }}" method="POST">
+                @csrf
+                <b-field label=@lang('users.fields.name') type="{{ $errors->has('name') ? 'is-danger' : null }}" message="{{ $errors->first('name') }}">
+                    <b-input type="text" name="name" id="name" value="{{ old('name') }}" maxlength="30"  required></b-input>
```
Las validaciones permiten hasta 255 caracteres, pero el HTML está limitando a 30

jortiz3109 commented on resources/views/register/create.blade.php:
```diff
@@ -0,0 +1,28 @@
+                </b-field>
+
+                <b-field label=@lang('users.fields.email') type="{{ $errors->has('email') ? 'is-danger' : null }}" message="{{ $errors->first('email') }}">
+                    <b-input type="email" name="email" id="email" value="{{ old('email') }}" maxlength="30" icon="account-lock"  required></b-input>
```
Las validaciones permiten hasta 255 caracteres, pero el HTML está limitando a 30

jortiz3109 commented on resources/views/register/create.blade.php:
```diff
@@ -0,0 +1,28 @@
+                </b-field>
+
+                <b-field label=@lang('users.fields.password') type="{{ $errors->has('password') ? 'is-danger' : null }}" message="{{ $errors->first('password') }}">
+                    <b-input type="password" name="password" id="password" value="{{ old('password') }}" maxlength="30" icon="lock" required></b-input>
```
Las validaciones permiten hasta 255 caracteres, pero el HTML está limitando a 30

jortiz3109 commented on resources/views/register/create.blade.php:
```diff
@@ -0,0 +1,28 @@
+                </b-field>
+
+                <b-field label=@lang('users.fields.password_confirmation')  type="{{ $errors->has('password_confirmation') ? 'is-danger' : null }}" message="{{ $errors->first('password_confirmation') }}">
+                    <b-input type="password" name="password_confirmation" id="password_confirmation" value="{{ old('password_confirmation') }}" maxlength="30" icon="lock" required></b-input>
```
Las validaciones permiten hasta 255 caracteres, pero el HTML está limitando a 30

jortiz3109 commented on routes/web.php:
```diff
@@ -17,6 +18,9 @@
 
     Route::name('logins.index')->get('/logins', LoginLogController::class);
+
+    Route::name('users.create')->get('/register', [UserController::class, 'create']);
```
El path de la ruta debería ser: /users/create

jortiz3109 commented on routes/web.php:
```diff
@@ -17,6 +18,9 @@
     Route::name('logins.index')->get('/logins', LoginLogController::class);
+
+    Route::name('users.create')->get('/register', [UserController::class, 'create']);
+    Route::name('users.store')->post('/register', [UserController::class, 'store']);
```
El path de la ruta debería ser /users

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,137 @@
+        $data = $this->developerData();
+        $response = $this->actingAs($user)->post('register',$data);
+
+        $response->assertRedirect('dashboard');
```
¿Esta prueba realmente "prueba" que el usuario fue creado en DB?

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,137 @@
+
+    }
+
+    public function test_required_name_user_register(): void
```
Estas 3 pruebas se pueden unificar en una sola usando un Data Provider
https://github.com/jortiz3109/escuela-php-workshop-1/blob/main/tests/Feature/API/Developers/StoreTest.php

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,137 @@
+        
+    }
+
+    public function test_create_date_user_register(): void
```
Yo no agregaría esta prueba, pues es un comportamiento del framework que ya está probado

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,137 @@
+        
+    }
+
+    public function test_create_id_user_register(): void
```
Creo que este podría ser un enfoque interesante para esta prueba.

```php
$user = User::factory()->create();
$this->actingAs($user)->post('/users', $this->developerData());
$this->assertEquals(User::latest()->first()->created_by, $user->id);
```

jortiz3109 commented on tests/Feature/Register/RegisterUserTest.php:
```diff
@@ -0,0 +1,137 @@
+        ]);        
+    }
+
+    public function developerData(): array
```
No lo sé Rick... se me hace como falso.

jortiz3109 reviewed (changes requested):
Todavía hay cosillas por mejorar, pero vamos bien

jeantech-zz committed:
feat: create_user_actions

jeantech-zz committed:
feat: rename view user

jeantech-zz committed:
fix: migration add user

jeantech-zz committed:
fix; request users

jeantech-zz committed:
fix seeder Environment Variables

jeantech-zz committed:
fix translate buttons

jeantech-zz committed:
fix: large input form

jeantech-zz committed:
fix: name router user

jeantech-zz committed:
fix: rename route users

jeantech-zz committed:
fix test

juancolo committed:
test: fix test for user create and store

it should have a refactor into the distributtion of test. It is because it is mixing create and store tests.

juancolo committed:
refactor:  UserController refactor

juancolo committed:
refactor: web routes for user were refactor

juancolo committed:
test: add user data providers for testing

juancolo committed:
style: php cs fixer were ran

juancolo committed:
style: php cs fixer were ran

juancolo committed:
feat: add user enable action for check the status of email verified

juancolo committed:
style: php cs fixer were ran

juancolo committed:
fix: user views folder to scafolding convention

juancolo committed:
test: refactor dataProvider scafolding

juancolo committed:
test: User unit testing to increase the coverage

juancolo committed:
style: php cs fixer were ran

asilvabe commented on app/Actions/User/UserEnableDisableAction.php:
```diff
@@ -0,0 +1,17 @@
+
+use App\Models\User;
+
+class UserEnableDisableAction
```
Cambiar el nombre por `UserVerificationAction`

juancolo replied:
corregido

asilvabe commented on app/Actions/User/UserStoreAction.php:
```diff
@@ -0,0 +1,30 @@
+
+class UserStoreAction
+{
+    public static function execute(array $data, User $user): User
```
Usar esta acción únicamente para crear un nuevo usuario.
Implica que el usuario no debe llegar como un parámetro.
Revisar la lógica de implementación

juancolo replied:
corregido

asilvabe commented on app/Actions/User/UserStoreAction.php:
```diff
@@ -0,0 +1,30 @@
+
+    public static function isSaved(User $user): bool
+    {
+        $user->save() ? event(new Registered($user)) : Log::error(trans(''));
```
¿este evento si debe ir aquí?
Este evento debería generarse sólo cuando el usuario de registra a sí mismo

juancolo replied:
corregido

asilvabe commented on app/Actions/User/UserStoreAction.php:
```diff
@@ -0,0 +1,30 @@
+
+class UserStoreAction
+{
+    public static function execute(array $data, User $user): User
```
¿Este método debería ser estático?
¿no es mejor inyectar las dependencias a través del constructor?

asilvabe commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,24 @@
+
+    public function store(UserCreateRequest $request): RedirectResponse
+    {
+        UserStoreAction::execute($request->validated(), new User());
```
Recuerda que el usuario no debe pasarse por aquí, se crea en el action

juancolo replied:
corregido

asilvabe commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,24 @@
+{
+    public function create(CreateViewModel $viewModel): View
+    {
+        return view('users.create', $viewModel->toArray());
```
Quitar el `toArray`

juancolo replied:
corregido

asilvabe commented on app/Http/Requests/Users/UserCreateRequest.php:
```diff
@@ -0,0 +1,24 @@
+        return [
+            'name' => ['required', 'string', 'max:255', 'regex:/^[a-zA-Z\s]+$/u'],
+            'email' => ['required', 'email', 'max:255',  Rule::unique('users', 'email')->ignore($this->users)],
+            'password' => ['required', 'confirmed', Rules\Password::defaults()],
```
Importar la clase Password

juancolo replied:
corregido

asilvabe commented on app/Http/Requests/Users/UserCreateRequest.php:
```diff
@@ -0,0 +1,24 @@
+    {
+        return [
+            'name' => ['required', 'string', 'max:255', 'regex:/^[a-zA-Z\s]+$/u'],
+            'email' => ['required', 'email', 'max:255',  Rule::unique('users', 'email')->ignore($this->users)],
```
El `ignore` no debe aplicarse por tratarse de la creación del usuario

juancolo replied:
corregido

asilvabe commented on app/Http/Requests/Users/UserCreateRequest.php:
```diff
@@ -0,0 +1,24 @@
+    public function rules(): array
+    {
+        return [
+            'name' => ['required', 'string', 'max:255', 'regex:/^[a-zA-Z\s]+$/u'],
```
Revisar si existe una regla para esto y evitar utilizar la expresión regular

juancolo replied:
No hay una regla para letras y espacios

asilvabe commented on app/Models/User.php:
```diff
@@ -51,4 +54,9 @@ public function markAsDisabled(): void
         $this->save();
     }
+
+    public function isEmailVerified(self $user): bool
```
Elminar este parámetro, usarlo son la variable `$this`

juancolo replied:
corregido

asilvabe commented on resources/views/layouts/admin.blade.php:
```diff
@@ -26,7 +26,7 @@
                         <ul class="menu-list">
                             <li><a href="{{ route('permissions.index') }}"><em class="pr-2 mdi mdi-shield-lock"></em>@lang('permissions.navbar.title')</a></li>
-                            <li><a href="#"><em class="pr-2 mdi mdi-account-multiple"></em>@lang('Users')</a></li>
+                            <li><a href="#"><em class="pr-2 mdi mdi-account-multiple"></em></a></li>
```
¿en serio debía eliminarse el link de usuarios?

juancolo replied:
corregido

asilvabe commented on resources/views/users/create.blade.php:
```diff
@@ -0,0 +1,28 @@
+                </button>
+            </form>
+
+@endsection
```
Agregar la línea adiciona

asilvabe commented on resources/views/users/create.blade.php:
```diff
@@ -0,0 +1,28 @@
+                </b-field>
+
+                <button type="submit" class="button is-primary is-fullwidth"> <b-icon pack="fas" icon="save"></b-icon>&nbsp;
+                    @lang('users.buttons.save')
```
Reemplazar todos los usos de @lang por `{{ trans() }}`

juancolo replied:
corregido

asilvabe commented on resources/views/users/create.blade.php:
```diff
@@ -0,0 +1,28 @@
+                    <b-input type="password" name="password_confirmation" id="password_confirmation" value="{{ old('password_confirmation') }}" maxlength="255" icon="lock" required></b-input>
+                </b-field>
+
+                <button type="submit" class="button is-primary is-fullwidth"> <b-icon pack="fas" icon="save"></b-icon>&nbsp;
```
Agregar desde el viewModel con el array de botones

juancolo replied:
realizado

asilvabe commented on routes/web.php:
```diff
@@ -17,6 +18,8 @@
 
     Route::name('logins.index')->get('/logins', LoginLogController::class);
+
+    Route::resource('users', UserController::class);
```
Agregar `only('create')`

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+    /**
+     * @test
+     */
+    public function authenticated_user_can_view_the_create_view(): void
```
Reemplazar el comentario por el prefijo `test_`

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+    {
+        $user = User::factory()->create();
+        $this->actingAs($user)->get('/users/create')
+            ->assertStatus(Response::HTTP_OK);
```
Agregar la asersión que afirma que es una vista (assertViewIs)

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+    public function unauthenticated_user_can_not_view_the_create_view(): void
+    {
+        $this->get('/users/create')
+            ->assertStatus(Response::HTTP_FOUND);
```
Debe redirigir a la ruta login

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+        $data['updated_by'] = $user->id;
+
+        $this->actingAs($user)->post('/users', $data)
+            ->assertStatus(302);
```
Usar la constante de la clase Response

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+        $user = User::factory()->create();
+
+        $data['created_by'] = $user->id;
+        $data['updated_by'] = $user->id;
```
No debe asignarse estos valores aquí, ya que se asignar en la acción

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+     */
+    public function authenticated_user_can_view_the_create_view(): void
+    {
+        $user = User::factory()->create();
```
Reemplazar por uso del Trait en toda la clase

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+            'name' => $data['name'],
+            'email' => $data['email'],
+            'created_by' => $user->id,
+            'updated_by' => $user->id,
```
`updated_by` no debe ir aquí

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+        $user = User::factory()->create();
+
+        $this->actingAs($user)->post('/users', $data)
+            ->assertStatus(302)
```
Usar la constante de `Response`

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+     * @param string $field
+     * @test
+     */
+    public function a_user_store_required_data(array $data, string $field): void
```
Buscar un mejor nombre: Ej. `test_can_not_create_an_user_with_invaid_data`

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    /**
+     * @dataProvider validUserStatusData
```
¿si es necesario este provider?

juancolo replied:
Corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+     * @param array $status
+     * @test
+     */
+    public function disabled_user_register(array $data, array $status): void
```
Pensar en un nombre más descriptivo

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    /**
+     * @dataProvider validUserData
```
No es necesario este provider

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+        $data['updated_by'] = $user->id;
+
+        $this->actingAs($user)->post('/users', $data)
+            ->assertRedirect('/dashboard');
```
¿Si debería redirigir a esta ruta?

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+     * @param array $data
+     * @test
+     */
+    public function create_id_user_register(array $data): void
```
Pensar en un nombre mejor

juancolo replied:
corregido

asilvabe commented on tests/Feature/Users/CreateUserTest.php:
```diff
@@ -0,0 +1,136 @@
+     * @param array $data
+     * @test
+     */
+    public function unique_email_user_register(array $data): void
```
Refactorizar implementando esto en la prueba de validaciones
Recordar la discusión entre Lauana y Alexandre sobre los provider's
Si lo dejas ocmo un test separado eliminar el uso del provider

juancolo replied:
corregido

asilvabe commented on tests/Support/User/UserFeatureDataProvider.php:
```diff
@@ -0,0 +1,68 @@
+            'empty password for user creation' => [
+                'data' => $this->getAttributes($this->name, $this->email, $this->invalid),
+                'field' => 'password',
+            ],
```
Incluir las reglas de validación que faltan

juancolo replied:
corregido

asilvabe commented on tests/Support/User/UserFeatureDataProvider.php:
```diff
@@ -0,0 +1,68 @@
+        ];
+    }
+
+    public function invalidUserData(): array
```
Revisar [este código](https://github.com/jortiz3109/simple-scrum-manager/blob/main/tests/DataProviders/Developer/StoreOrUpdaterDataProvider.php) para inspiración

juancolo replied:
corregido

asilvabe commented on tests/Support/User/UserUnitDataProvider.php:
```diff
@@ -0,0 +1,34 @@
+<?php
```
Eliminar este archivo

juancolo replied:
corregido

asilvabe commented on tests/Unit/Models/UserTest.php:
```diff
@@ -0,0 +1,39 @@
+<?php
```
Eliminar este archivo

juancolo replied:
corregido

juancolo committed:
Merge branch 'develop' into feature/user_register

juancolo committed:
fix: detele unused files

juancolo committed:
test: add and fix test for User  create and store Test

juancolo committed:
fix: take the pr comments and modify the execute metodh

juancolo committed:
fix: modify the create and store functions for UserController and routes

juancolo committed:
feat: create the UserVerificarionAction Class

juancolo committed:
feat: add user stored event and  sendUserEmailVerification listener

juancolo committed:
feat: add UserStored and SendUserEmailVerification in the EventServiceProvider

juancolo committed:
fix: fix based on pr comments

Fix User Model, users translations and UserCreateRequest

juancolo committed:
fix: Rename and change UserCreateViewModel

juancolo committed:
fix: usage of UserCreateViewModel

juancolo committed:
style: php cs fixer were run

juancolo committed:
fix: delete UserTest

juancolo committed:
test: fix the view asser in CreateTest

juancolo committed:
test: add UserTest for coverage

juancolo committed:
style: run php-cs-fixer

juancolo committed:
test: add test to User and fix external used clases

juancolo committed:
test: add test to UserTest for change enabled status

juancolo committed:
merge: pull from develop

juancolo committed:
merge: pull from develop

juancolo committed:
fix: remove unused variable

juancolo committed:
test: fix sonar test code smells

juancolo committed:
style: php-cs-fixer

juancolo committed:
Merge branch 'develop' of https://github.com/jortiz3109/escuela-php-panel into feature/users-registrer

juancolo committed:
fix: user controller miss }

juancolo committed:
style: php-cs-fixer

juancolo committed:
fix: routes for create and store were delted

juancolo committed:
test: fix the email of HasAuthenticadUSer

asilvabe committed:
Merge branch 'develop' into feature/users-registrer

asilvabe committed:
feat: remove duplicated lang file keys

juancolo committed:
test: fix User StoreTest

juancolo committed:
fix: User Email validiation fixes and UserStore route fixes into ViewModel

juancolo committed:
feat: add button create into user index view

juancolo committed:
Merge branch 'develop' of https://github.com/jortiz3109/escuela-php-panel into feature/users-registrer

juancolo committed:
Merge branch 'feature/users-registrer' of https://github.com/jortiz3109/escuela-php-panel into feature/users-registrer

juancolo committed:
fix: miss method deleted

juancolo committed:
fix: duplications in users.index replaced with a constant

juancolo committed:
fix: replace user.index text for a constant
